File: python/src/behavioral/observer/concrete_subject.py
from .subject import Subject
from .observer import Observer
from typing import Optional


class ConcreteSubject(Subject):
    """
    Concrete Subject stores state of interest and sends notifications
    when its state changes.
    """

    def __init__(self):
        self._observers: list[Observer] = []
        self._state: str | None = None

    def attach(self, observer: Observer) -> None:
        if observer not in self._observers:
            self._observers.append(observer)

    def detach(self, observer: Observer) -> None:
        if observer in self._observers:
            self._observers.remove(observer)

    def notify(self, message: str) -> None:
        for observer in self._observers:
            observer.update(message)

    def set_state(self, new_state: str) -> None:
        self._state = new_state
        self.notify(f"State changed to: {new_state}")

    def get_state(self) -> Optional[str]:
        return self._state


# Example Usage
#
# from .concrete_subject import ConcreteSubject
# from .concrete_observer_a import ConcreteObserverA
# from .concrete_observer_b import ConcreteObserverB
#
# subject = ConcreteSubject()
#
# observer_a = ConcreteObserverA()
# observer_b = ConcreteObserverB()
#
# subject.attach(observer_a)
# subject.attach(observer_b)
#
# subject.set_state("running")
#
# print(observer_a.last_message)
# # Observer A received: State changed to: running
#
# print(observer_b.last_message)
# # Observer B processed: STATE CHANGED TO: RUNNING
#
#
# Notes:
# - Observers are notified immediately when subject state changes.
# - You can detach observers at any time using subject.detach(observer).
# - The subject does not need to know the concrete observer types.

File: go/src/behavioral/observer/concrete_subject.go
package observer

import "sync"

type ConcreteSubject struct {
	mu        sync.RWMutex
	state     string
	observers map[string]Observer
}

func NewConcreteSubject() *ConcreteSubject {
	return &ConcreteSubject{
		observers: make(map[string]Observer),
	}
}

func (s *ConcreteSubject) Attach(o Observer) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.observers[o.ID()] = o
}

func (s *ConcreteSubject) Detach(id string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.observers, id)
}

func (s *ConcreteSubject) Notify() {
	// Copy observers under read lock, then call Update() outside the lock.
	// This avoids deadlocks if observers call back into the subject.
	s.mu.RLock()
	state := s.state
	copied := make([]Observer, 0, len(s.observers))
	for _, o := range s.observers {
		copied = append(copied, o)
	}
	s.mu.RUnlock()

	for _, o := range copied {
		o.Update(state)
	}
}

func (s *ConcreteSubject) SetState(state string) {
	s.mu.Lock()
	s.state = state
	s.mu.Unlock()
	s.Notify()
}

func (s *ConcreteSubject) GetState() string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.state
}
